chore: check server version on sign-in and launch

deansheather opened the pull request:
Updates CredentialModel to have an additional state `Unknown` during startup while credentials are validated in the background asynchronously (15s timeout).

While loading credentials on startup, the tray app shows a loading message.

While updating credentials, we now check that the server version falls in our expected range.

The sign in page now uses a WinUI 3 Dialog to show errors.

Closes #42

ethanndickson commented on App/ViewModels/SignInViewModel.cs:
```diff
@@ -126,7 +126,14 @@ public async Task TokenPage_SignIn(SignInWindow signInWindow)
+                CloseButtonText = "Ok",
+                XamlRoot = signInWindow.Content.XamlRoot,
+            };
+            _ = await dialog.ShowAsync();
```
On the Mac app, I made the same decision as it was easier to produce a variable size popup with the error message than to either:
- Reserve an arbitrary amount of space on the window for an error message of an unknown length.
- Resize the window to accommodate the size of the error message.

Though maybe we revisit that. We could display a short error message on the window that can be clicked for details, something along those lines?

deansheather replied:
I'm pretty sure this is very common design language to use a popup dialog for these types of things. The WinUI3 Dialog docs page even mentions that it's used for errors.

But yeah it's pretty much because of what Ethan said.

deansheather commented on App/Models/CredentialModel.cs:
```diff
@@ -2,24 +2,33 @@ namespace Coder.Desktop.App.Models;
     public string? CoderUrl { get; set; }
     public string? ApiToken { get; set; }
 
+    public string? Username { get; set; }
```
I made them `init` instead of `set`, which means they can only be set at instantiation. Essentially the same as `readonly` but allows the `{ }` syntax I use below

deansheather commented:
@spikecurtis I pretty heavily reworked CredentialManager:

- Duplicate `LoadCredentials` calls now get merged (the same Task will be returned from both, so they will receive identical exceptions/results)
- A `SetCredentials` call during an ongoing `LoadCredentials` call will cancel it
- Added tests
    - To add tests I needed to add `ICoderApiClientFactory` and `ICoderApiClient` for mocking purposes
    - Pulled the `NativeApi` components of `CredentialManager` into a new type `WindowsCredentialBackend` (and added interface `ICredentialBackend` for mocking)

deansheather commented:
To protect things from concurrent access:
- `SetCredentials` holds `_opLock` for the entire operation
- `LoadCredentials` will hold `_opLock` while it creates the "inner" task then releases it
    - The "inner" task will grab the lock again when it's ready to write the loaded value

spikecurtis reviewed (commented):
I like the new CredentialManager overall, but still a bit concerned about invoking callbacks while holding the lock. A couple other small nits FYI.

deansheather commented on App/Services/CredentialManager.cs:
```diff
@@ -18,119 +18,286 @@ public class RawCredentials
+    private void UpdateState(CredentialModel newModel)
+    {
+        _latestCredentials = newModel;
         CredentialsChanged?.Invoke(this, newModel.Clone());
```
If I called this in a background Task instead would it be better? It means it wouldn't block either which would be nice.

deansheather replied:
Switched this to be in a background task

deansheather commented on App/App.xaml.cs:
```diff
@@ -45,17 +48,29 @@ public async Task ExitApplication()
+        // with no cancellation, the method itself will impose a timeout on the
+        // HTTP portion.
+        var credentialManager = _services.GetRequiredService<ICredentialManager>();
+        _ = credentialManager.LoadCredentials(CancellationToken.None);
```
Yep, we run it in the background and ignore all errors. If the credentials fail to load, the state will transition to "invalid" which will show the "please sign in" screen.

deansheather commented on App/Services/CredentialManager.cs:
```diff
@@ -18,119 +18,286 @@ public class RawCredentials
+        }
+        catch (Exception e)
+        {
+            throw new InvalidOperationException("Could not connect to or verify Coder server", e);
```
I added parsing for `codersdk.Response` and a new exception type `CoderApiHttpException`
![image](https://github.com/user-attachments/assets/e7118ca4-ae9f-4c2a-9578-09c885795760)
